### ISSUE TITLE
Need to reduce Keras warnings

### ISSUE THREAD

**Issue by cmdupuis3** [xarray-contrib/xbatcher#38]
In trying to solve #36, I found that this warning

`WARNING:tensorflow:Keras is training/fitting/evaluating on array-like data. Keras may not be optimized for this format, so if your input data format is supported by TensorFlow I/O (https://github.com/tensorflow/io) we recommend using that to load a Dataset instead.`

is duplicated so many times in my Jupyter notebook that it explodes my browser's RAM usage and freezes my computer! Recommend only printing unique warnings, but no clue how to implement that. Setting `model.fit` option `verbose=0` suppresses most of the spam, but this warning gets through :/ I assume that the solution to this would also make setting `verbose=0` less important.

**Comment by rabernat** [comment]
If these are normal python warnings, you should be able to suppress them as described here: https://docs.python.org/3/library/warnings.html#overriding-the-default-filter

**Comment by cmdupuis3** [comment]
No luck so far.

**Comment by cmdupuis3** [comment]
This style of warning suppression seems promising, but it's not working:

`import warnings

def fxn():
    warnings.warn(Warning)

with warnings.catch_warnings():
    warnings.simplefilter("ignore")
    fxn()`

**Comment by cmdupuis3** [comment]
Adding `clear_output(wait=True)` in the `for batch in bgen` loop helps. It would make sense to add this to the documentation, I expect this will be a common problem. I'll see if I can add it in the (future) pull request for #36

**Comment by jhamman** [comment]
@cmdupuis3 - how would I reproduce the warning you are getting? Can you share a contained example?

**Comment by cmdupuis3** [comment]
@jhamman I don't have a simple example of this handy, can I send you a gist of what I have?

**Comment by cmdupuis3** [comment]
I think I have something workable...

Using my update to xbatcher [here](https://github.com/cmdupuis3/xbatcher/tree/batch_dim), you can run this:

    import numpy as np
    import xarray as xr
    
    from IPython.display import clear_output
    import tensorflow as tf
    import gc

    def filter_none(xs):
        return list(filter(lambda x: x is not None, xs))

    import importlib.util
    spec = importlib.util.spec_from_file_location("xbatcher", "<your xbatcher path>")
    xb = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(xb)

    from dataclasses import dataclass
    from typing import Iterable
    
    @dataclass
    class Scenario:
        vanilla: Iterable[str]
        stencil_2D: Iterable[str]
        stencil_3D: Iterable[str]
        target: Iterable[str]
        name: str
        
    sc = Scenario(['X','Y'], ['Z'], [], ['t1', 't2'], name='stuff')
    
    X = xr.DataArray(np.random.rand(640, 640), dims=['nlon', 'nlat'], name="X")
    Y = xr.DataArray(np.random.rand(640, 640), dims=['nlon', 'nlat'], name="Y")
    Z = xr.DataArray(np.random.rand(640, 640), dims=['nlon', 'nlat'], name="Z")
    
    t1 = xr.DataArray(np.random.rand(640, 640), dims=['nlon', 'nlat'], name="t1")
    t2 = xr.DataArray(np.random.rand(640, 640), dims=['nlon', 'nlat'], name="t2")
    
    ds = xr.Dataset({'X':X, 'Y':Y, 'Z':Z, 't1':t1, 't2':t2})
    
    def train(ds, sc, conv_dims_2D = [20,20], conv_dims_3D = [3,20,20], nfilters=80):
    
        nlons = conv_dims_2D[0]
        nlats = conv_dims_2D[1]
        
        bgen = xb.BatchGenerator(
            ds,
            {'nlon':nlons,        'nlat':nlats},
            {'nlon':int(nlons/2), 'nlat':int(nlats/2)}
        )
    
        input_vanilla = tf.keras.Input(shape=len(sc.vanilla))
        
        input_stencil_2D = None
        input_stencil_3D = None
        reshape_layer_2D = None
        reshape_layer_3D = None
        if(sc.stencil_2D):
            input_stencil_2D = tf.keras.Input(shape=tuple(conv_dims_2D) + (len(sc.stencil_2D),))
            conv_layer_2D    = tf.keras.layers.Conv2D(nfilters, conv_dims_2D)(input_stencil_2D)
            reshape_layer_2D = tf.keras.layers.Reshape((nfilters,))(conv_layer_2D)
        if(sc.stencil_3D):
            input_stencil_3D = tf.keras.Input(shape=tuple(conv_dims_3D) + (len(sc.stencil_3D),))
            conv_layer_3D    = tf.keras.layers.Conv3D(nfilters, conv_dims_3D)(input_stencil_3D)
            reshape_layer_3D = tf.keras.layers.Reshape((nfilters,))(conv_layer_3D)
        
        concats = filter_none([reshape_layer_2D, reshape_layer_3D, input_vanilla])
        concat_layer = tf.keras.layers.concatenate(concats)
        combined_relu = tf.keras.layers.LeakyReLU(alpha=0.3)(concat_layer)
        hidden_layer_1 = tf.keras.layers.Dense(40, activation='relu')(combined_relu)
        hidden_layer_2 = tf.keras.layers.Dense(20, activation='relu')(hidden_layer_1)
        hidden_layer_3 = tf.keras.layers.Dense(10, activation='relu')(hidden_layer_2)
        output_layer = tf.keras.layers.Dense(2)(hidden_layer_3)
    
        model = tf.keras.Model(inputs=filter_none([input_stencil_2D, input_stencil_3D, input_vanilla]), outputs=output_layer)
        model.compile(loss='mae', optimizer='Adam', metrics=['mae', 'mse', 'accuracy'])
        
        for batch in bgen:
            batch_stencil_2D = None
            batch_stencil_3D = None
            if(sc.stencil_2D):
                batch_stencil_2D = xr.concat([batch[x] for x in sc.stencil_2D], dim='var').transpose(...,'var')
            if(sc.stencil_3D):
                batch_stencil_3D = xr.concat([batch[x] for x in sc.stencil_3D], dim='var').transpose(...,'var')
    
            batch_vanilla = xr.concat([batch[x].isel(nlat=int(nlats/2), nlon=int(nlons/2)) for x in sc.vanilla], dim='var').transpose(...,'var')
            batch_target  = xr.concat([batch[x].isel(nlat=int(nlats/2), nlon=int(nlons/2)) for x in  sc.target], dim='var').transpose(...,'var')
            
            model.fit(filter_none([batch_stencil_2D, batch_stencil_3D, batch_vanilla]),
                      batch_target,
                      batch_size=32, epochs=2, verbose=0)
            clear_output(wait=True)
    
        model.save('models/'+ sc.name)
    
        return model
        
    train(ds, sc)

**Comment by cmdupuis3** [comment]
Looking back at this, I think this issue is more a consequence of iterating over separate batches, rather than a problem with xbatcher itself. I could see a solution here being useful in the context of debugging an ML model, but I don't think it would make sense as part of xbatcher's cor functionality.

Closing for now.
